### ISSUE TITLE
SequentialSampler Not as Intended..

### ISSUE THREAD

**Issue by sajaddarabi** [pytorch/pytorch#23470]
## 🐛 Bug

This pesky bug was quite an annoying one! I had train set leak into my test set as the SequentialSampler is not implemented correctly I am presuming. 

Referring to [here](https://pytorch.org/docs/stable/_modules/torch/utils/data/sampler.html), the `SequentialSampler(Sampler)` is supposed to return elements from the dataset in order, instead it's returning an index starting from 0. 

What would be the point of passing a dataset to this function then?

## To Reproduce

```
import numpy as np
from torch.utils.data.sampler import SequentialSampler

split = 0.25
n_samples = 1000
full_idx = np.arange(0, n_samples)

val_idx = full_idx[:int(split * n_samples)]
train_idx = np.delete(full_idx, np.arange(0, int(split * n_samples)))

val_sampler = SequentialSampler(val_idx)
train_sampler = SequentialSampler(train_idx)

print([i for i in iter(train_sampler)]) # will print 0 to 750
print([i for i in iter(val_sampler)])  # will print 0 to 249

```

## Expected behavior

The expected behavior in the above is to have sequential sampler return the samples sequentially contained in the dataset passed to the sampler..

i.e. the print lines should print out 250 - 1000 and 0 - 249 respectively.

A correct implementation could be:

```
class SequentialSampler(Sampler):
    r"""Samples elements sequentially, always in the same order.

    Arguments:
        data_source (Dataset): dataset to sample from
    """

    def __init__(self, data_source):
        self.data_source = data_source

    def __iter__(self):
        return (self.data_source[i] for i in range(len(self.data_source)))

    def __len__(self):
        return len(self.data_source)
```

## Environment

Collecting environment information...
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 410.104
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.5.1
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn.so.7

Versions of relevant libraries:
[pip] numpy==1.15.4
[pip] pytorch-memlab==0.0.3
[pip] torch==1.0.0
[pip] torchvision==0.2.1
[conda] pytorch-memlab            0.0.3                     <pip>
[conda] torch                     1.0.0                     <pip>
[conda] torchvision               0.2.1                     <pip>

**Comment by SsnL** [comment]
No, SequentialSampler works as intended. It takes in the dataset, not the set of indices. Your use case just needs another sampler.

**Comment by sajaddarabi** [comment]
You're right. 

It must have misinterpreted "Samples elements sequentially, always in the same order."... to mean it returns items from the inputted dataset sequentially as opposed to providing a set of indexes that can be used to sample from the dataset sequentially.
